### PR TITLE
Bed piston fix

### LINKED ISSUE #3954: Pushing a bed with a piston crashes server
Client version:   1.12.1
Server OS:        Linux
Commit id:        ad3192d6962536e21233710febca9f128b8e0b80


### Expected behavior
A bed pushed by a piston should convert to a pickup.

### Actual behavior
The server crashes.
```
[17:39:00]   D:    | Cuberite has encountered an error and needs to close                                                                                                                                                                     
[17:39:00] Details | SIGSEGV: Segmentation fault                                                                                                                                                                                              
[17:39:00] Cuberite local build build id: Unknown                                                                                                                                                                                             
[17:39:00] from commit id: approx: ad3192d6962536e21233710febca9f128b8e0b80 built at: approx: 2017-08-21 16:00:49 +0200                                                                                                                       
./Cuberite(_Z15PrintStackTracev+0x15)[0x55b8502c1435]                                                                                                                                                                                         
./Cuberite(+0x2a6424)[0x55b8502af424]                                                                                                                                                                                                         
/lib/x86_64-linux-gnu/libc.so.6(+0x33030)[0x7efd34c4f030]                                                                                                                                                                                     
./Cuberite(_ZN16cBlockBedHandler16ConvertToPickupsEP7cEntityR6cItemshiii+0x2f)[0x55b8503cd90f]                                                                                                                                                
./Cuberite(_ZN13cBlockHandler9DropBlockER15cChunkInterfaceR15cWorldInterfaceR21cBlockPluginInterfaceP7cEntityiiib+0x376)[0x55b8503b6456]                                                                                                      
./Cuberite(_ZN19cBlockPistonHandler10PushBlocksERKSt13unordered_setI7Vector3IiE12VectorHasherIiESt8equal_toIS2_ESaIS2_EER6cWorldRKS2_+0x484)[0x55b8503ca5d4]                                                                                  
./Cuberite(_ZN19cBlockPistonHandler12ExtendPistonE7Vector3IiER6cWorld+0x27b)[0x55b8503cc39b]                                                                                                                                                  
./Cuberite(_ZNK14cPistonHandler6UpdateER6cWorldRK7Vector3IiEhhN16cRedstoneHandler12PoweringDataE+0x128)[0x55b8505468c8]                                                                                                                       
./Cuberite(_ZN29cIncrementalRedstoneSimulator8SimulateEf+0x2e8)[0x55b85053dac8]                                                                                                                                                               
./Cuberite(_ZN17cSimulatorManager8SimulateEf+0x5e)[0x55b85053b37e]                                                                                                                                                                            
./Cuberite(_ZN6cWorld4TickENSt6chrono8durationIlSt5ratioILl1ELl1000EEEES4_+0x39f)[0x55b8502a460f]                                                                                                                                             
./Cuberite(_ZN6cWorld11cTickThread7ExecuteEv+0x84)[0x55b8502a48f4]                                                                                                                                                                            
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xb9e6f)[0x7efd357ace6f]                                                                                                                                                                            
/lib/x86_64-linux-gnu/libpthread.so.0(+0x7494)[0x7efd34fc2494]                                                                                                                                                                                
/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7efd34d04aff]                                                                                                                                                                                   
[17:39:00]   D:    | Cuberite has encountered an error and needs to close                                                                                                                                                                     
[17:39:00] Details | SIGABRT: Server self-terminated due to an internal fault                                                                                                                                                                 
[17:39:00] Cuberite local build build id: Unknown                                                                                                                                                                                             
[17:39:00] from commit id: approx: ad3192d6962536e21233710febca9f128b8e0b80 built at: approx: 2017-08-21 16:00:49 +0200                                                                                                                       
./Cuberite(_Z15PrintStackTracev+0x15)[0x55b8502c1435]                                                                                                                                                                                         
./Cuberite(+0x2a6424)[0x55b8502af424]                                                                                                                                                                                                         
/lib/x86_64-linux-gnu/libc.so.6(+0x33030)[0x7efd34c4f030]                                                                                                                                                                                     
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcf)[0x7efd34c4efcf]                                                                                                                                                                                 
/lib/x86_64-linux-gnu/libc.so.6(abort+0x16a)[0x7efd34c503fa]                                                                                                                                                                                  
./Cuberite(+0x2a6429)[0x55b8502af429]
/lib/x86_64-linux-gnu/libc.so.6(+0x33030)[0x7efd34c4f030]                
[17:39:00] Kicking player Wither_Skeleton for "Server is restarting. Please wait..."
[17:39:00] Kicking player waeve for "Server is restarting. Please wait..."
./Cuberite(_ZN16cBlockBedHandler16ConvertToPickupsEP7cEntityR6cItemshiii+0x2f)[0x55b8503cd90f]
./Cuberite(_ZN13cBlockHandler9DropBlockER15cChunkInterfaceR15cWorldInterfaceR21cBlockPluginInterfaceP7cEntityiiib+0x376)[0x55b8503b6456]
./Cuberite(_ZN19cBlockPistonHandler10PushBlocksERKSt13unordered_setI7Vector3IiE12VectorHasherIiESt8equal_toIS2_ESaIS2_EER6cWorldRKS2_+0x484)[0x55b8503ca5d4]
./Cuberite(_ZN19cBlockPistonHandler12ExtendPistonE7Vector3IiER6cWorld+0x27b)[0x55b8503cc39b]
./Cuberite(_ZNK14cPistonHandler6UpdateER6cWorldRK7Vector3IiEhhN16cRedstoneHandler12PoweringDataE+0x128)[0x55b8505468c8]
./Cuberite(_ZN29cIncrementalRedstoneSimulator8SimulateEf+0x2e8)[0x55b85053dac8]
./Cuberite(_ZN17cSimulatorManager8SimulateEf+0x5e)[0x55b85053b37e]
./Cuberite(_ZN6cWorld4TickENSt6chrono8durationIlSt5ratioILl1ELl1000EEEES4_+0x39f)[0x55b8502a460f]
./Cuberite(_ZN6cWorld11cTickThread7ExecuteEv+0x84)[0x55b8502a48f4]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xb9e6f)[0x7efd357ace6f]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x7494)[0x7efd34fc2494]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7efd34d04aff]
```

### Steps to reproduce the behavior
Place a piston next to a bed and activate the piston.

### PR DESCRIPTION
**Pull request by lkolbly**
When a block is turned into drops by a piston, the a_Digger argument is nullptr, because no entity is the digger. Beds use the GetWorld() method of a_Digger to figure out what color they are.

This PR passes a cWorldInterface directly to the ConvertToPickups method. It also changes how piston's PushBlocks method sets the original block to air, to make sure the block still exists when DropBlock is called.

With credit to @Seadragon91, who figured out that pistons were clearing the block before the DropBlock method was called.

Fixes #3954.

### CODE REVIEW

File: src/Blocks/BlockBed.cpp [hunk 1]
```diff
@@ -209,7 +209,7 @@ void cBlockBedHandler::OnPlacedByPlayer(cChunkInterface & a_ChunkInterface, cWor
 
 
 
-void cBlockBedHandler::ConvertToPickups(cEntity * a_Digger, cItems & a_Pickups, NIBBLETYPE a_BlockMeta, int a_BlockX, int a_BlockY, int a_BlockZ)
+void cBlockBedHandler::ConvertToPickups(cWorldInterface & a_WorldInterface, cEntity * a_Digger, cItems & a_Pickups, NIBBLETYPE a_BlockMeta, int a_BlockX, int a_BlockY, int a_BlockZ)
```
Comment (Seadragon91) [inline src/Blocks/BlockBed.cpp]: Please remove `a_Digger` is not needed anymore

File: src/Blocks/BlockBed.cpp [hunk 2]
```diff
@@ -219,11 +219,12 @@ void cBlockBedHandler::ConvertToPickups(cEntity * a_Digger, cItems & a_Pickups,
 
 		virtual bool Item(cBedEntity * a_Bed) override
 		{
+			LOG("Got bed color");
```
Comment (Seadragon91) [inline src/Blocks/BlockBed.cpp]: Can be removed.

**Comment by Seadragon91** [comment]
A few changes . I tested it and it works good, thanks :)

File: src/Blocks/BlockBed.cpp (hunk 1 above)
Comment (lkolbly) [inline src/Blocks/BlockBed.cpp]: Removed. For implementing the Fortune enchantment, does it seem reasonable to add another argument "int a_FortuneLevel" here?

File: src/Blocks/BlockBed.cpp (hunk 2 above)
Comment (lkolbly) [inline src/Blocks/BlockBed.cpp]: Done.

File: src/Blocks/BlockHandler.h [hunk 3]
```diff
@@ -88,8 +88,8 @@ class cBlockHandler
 	virtual void ConvertToPickups(cItems & a_Pickups, NIBBLETYPE a_BlockMeta);
 
 	/** Called when the item is mined to convert it into pickups. Pickups may specify multiple items. Appends items to a_Pickups, preserves its original contents.
-	Overloaded method with coords and digger, for blocks that needs to access the block entity, e.g. a bed */
-	virtual void ConvertToPickups(cEntity * a_Digger, cItems & a_Pickups, NIBBLETYPE a_BlockMeta, int a_BlockX, int a_BlockY, int a_BlockZ) {}
+	Overloaded method with coords, digger, and world interface for blocks that needs to access the block entity, e.g. a bed. Note that a_Digger could be null e.g. when a piston causes the bed to drop */
```
Comment (Seadragon91) [inline src/Blocks/BlockHandler.h]: Sorry here also `digger`

Comment (lkolbly) [inline src/Blocks/BlockHandler.h]: Whoops, my bad. Fixed.

File: src/Blocks/BlockHandler.h [hunk 4]
```diff
@@ -88,8 +88,8 @@ class cBlockHandler
 	virtual void ConvertToPickups(cItems & a_Pickups, NIBBLETYPE a_BlockMeta);
 
 	/** Called when the item is mined to convert it into pickups. Pickups may specify multiple items. Appends items to a_Pickups, preserves its original contents.
-	Overloaded method with coords and digger, for blocks that needs to access the block entity, e.g. a bed */
-	virtual void ConvertToPickups(cEntity * a_Digger, cItems & a_Pickups, NIBBLETYPE a_BlockMeta, int a_BlockX, int a_BlockY, int a_BlockZ) {}
+	Overloaded method with coords and world interface for blocks that needs to access the block entity, e.g. a bed. Note that a_Digger could be null e.g. when a piston causes the bed to drop */
```
Comment (Seadragon91) [inline src/Blocks/BlockHandler.h]: `Note that a_Digger could be null e.g. when a piston causes the bed to drop` :)

**Comment by lkolbly** [comment]
Hm, what's the difference between DropBlock and ConvertToPickups? It looks to me like the overloaded ConvertToPickups method can be replaced by DropBlock entirely, because that's what BlockTallGrass does (it too has a complicated DropBlock logic, but it just overrides DropBlock).

**Comment by madmaxoft** [comment]
The new code should have used a `Vector3i` for the coords :(

**Pull request merged**